roles/sap_swpm/tasks/swpm/prepare_software.yml

```
@@ -12,49 +12,105 @@
-  ansible.builtin.file:
-    path: "{{ sap_swpm_software_path }}"
-    state: directory
-    recurse: yes
-    mode: '0755'
-    owner: root
-    group: root
+- name: SAP SWPM Pre Install - Set directory and file permissions
+  when: sap_swpm_set_file_permissions
+  block:
+
+    - name: SAP SWPM Pre Install - Find directories
+      ansible.builtin.find:
+        path: "{{ sap_swpm_software_path }}"
+        file_type: directory
+      register: __sap_swpm_register_find_result_directories
+
+    - name: SAP SWPM Pre Install - Find non-SAPCAR files
+      ansible.builtin.find:
+        path: "{{ sap_swpm_software_path }}"
+        file_type: file
+        recurse: true
+        excludes: "SAPCAR*EXE"
+      register: __sap_swpm_register_find_result_files_non_sapcar
+
+    - name: SAP SWPM Pre Install - Create list of absolute directory names from the find result
+      ansible.builtin.set_fact:
+        __sap_swpm_fact_directories: "{{ __sap_swpm_fact_directories | d([]) + [item.path] }}"
+      loop: "{{ __sap_swpm_register_find_result_directories.files }}"
+      when: __sap_swpm_register_find_result_directories is defined
+
+    - name: SAP SWPM Pre Install - Create list of absolute file names for non-SAPCAR files from the find result
+      ansible.builtin.set_fact:
+        __sap_swpm_fact_files_non_sapcar: "{{ __sap_swpm_fact_files_non_sapcar | d([]) + [item.path] }}"
+      loop: "{{ __sap_swpm_register_find_result_files_non_sapcar.files }}"
+      when: __sap_swpm_register_find_result_files_non_sapcar is defined
+
+    - name: SAP SWPM Pre Install - Ensure correct permissions and ownership of all directories
+      ansible.builtin.file:
+        path: "{{ line_item }}"
+        recurse: no
+        mode: "{{ sap_swpm_software_directory_mode }}"
+        owner: "{{ sap_swpm_software_directory_owner }}"
+        group: "{{ sap_swpm_software_directory_group }}"
+      loop: "{{ __sap_swpm_fact_directories }}"
+      loop_control:
+        loop_var: line_item
+      when:
+        - __sap_swpm_fact_directories is defined
+        - __sap_swpm_register_find_result_directories is defined
+
+    - name: SAP SWPM Pre Install - Create argument list for chown and chmod of non-SAPCAR*EXE files
+      ansible.builtin.set_fact:
+        __sap_swpm_fact_files_non_sapcar_chown_arg_list: "{{ __sap_swpm_fact_files_non_sapcar | map('quote') | join(' ') }}"
+
+# Reasons for noqa:
+# - command-instead-of-module: Shorter execution time compared to looping over a list when using the file module
+# - no-changed-when: Not worth checking permissions and ownership before this task and comparing afterwards
+    - name: SAP SWPM Pre Install - Ensure correct permissions and ownership of all non-SAPCAR files # noqa command-instead-of-module no-changed-when
+      ansible.builtin.shell: >
+         set -o pipefail && chown {{ sap_swpm_files_non_sapcar_owner }}:{{ sap_swpm_files_non_sapcar_group }} \
```

```suggestion
         chown {{ sap_swpm_files_non_sapcar_owner }}:{{ sap_swpm_files_non_sapcar_group }} \
```

There are no pipeline command so the set pipefail is not needed!

---

From looking at the code, I agree that it doesn't make much sense because there are no pipelines.
If I remember correctly, the task does not fail without pipefail when one of the commands (maybe the first one?) fails (maybe a side effect).
Maybe I should run some more tests and then either:
- document why it is useful, or
- remove it

---

if it fails there are something wrong in the ansible builtin shell module, it doesn't make sense in the scope of the content that set.

---

I tested and fixed it in https://github.com/sap-linuxlab/community.sap_install/pull/665/commits/538a41073b8b708f1003bdccdf0b7ad85790e734 . I could not find a problem with the shell command again.

---

In case that fails, we should consider use **pipelines** `||` instead of **logical and** `&&` to be more consistent in what we have inside that:
```
set -o pipefail 
chown {{ sap_swpm_files_non_sapcar_owner }}:{{ sap_swpm_files_non_sapcar_group }} \
{{ __sap_swpm_fact_files_non_sapcar_chown_arg_list }} || 
chmod {{ sap_swpm_files_non_sapcar_mode }} \
{{ __sap_swpm_fact_files_non_sapcar_chown_arg_list }}
```

---

Ok, so this is a resolved conversation!